Improve Decentralization Security Best Practices (#2238)
dfinity/portal: docs/developer-docs/security/rust-canister-development-security-best-practices.md

## Smart contracts canister control

### Use a decentralized governance system like SNS to put dapps under decentralized control

#### Security concern

> The controller of centralized components of a dapp (such as centrally controlled canisters or cloud services) can change or update these components whenever they like. If a canister or off-chain component e.g. holds assets such as ICP, ckBTC or ckETH on the user's behalf, this effectively means that the controller could decide at any time to steal these funds by e.g. updating the canister and transferring the assets to their account.

Reviewer: ```suggestion
The controller of centralized components of a dapp (such as centrally controlled canisters or cloud services) can change or update these components whenever they like. If a canister or off-chain component e.g. holds assets such as ICP, ckBTC or ckETH on a user's behalf, this effectively means that the controller could decide at any time to steal these funds by e.g. updating the canister and transferring the assets to their account.
```

> The controller of centralized components of a dapp (such as centrally controlled canisters or cloud services) can change or update these components whenever they like. If a canister or off-chain component e.g. holds assets such as ICP, ckBTC or ckETH on the user's behalf, this effectively means that the controller could decide at any time to steal these funds by e.g. updating the canister and transferring the assets to their account.

Author: @roelstorms  mentioned below that he is happy with the changes. Resolving this thread.

Furthermore, the controller of canisters serving web content (such as e.g. the [asset canister](../../references/asset-canister)) could maliciously modify the web application to e.g. steal user funds or perform security sensitive actions on the user's behalf. For example, if [Internet Identity](../../developer-docs/integrations/internet-identity/overview) is used, the user principal's private key for the given origin is stored in the browser storage, and a malicious app can therefore fully control the private key and the user's session and any assets controlled by that key.  

#### Recommendations

> The following list starts with recommendations for dapps under centralized control, and then provides increasingly stronger recommendations to strengthen the dapp's security in terms of decentralization. 

Reviewer: I stumbled over this a bit.. I guess we want to say 1) we start with a centralized setting and go to increasinly decetnralized ones and 2) the decentralized is more favorable from a security pov?

1. **The dapp uses central, off-chain components:** The application makes use of centralized components that are e.g. running in the cloud. The owners of these cloud services have full control over the application and assets managed by it. Your application should likely be further decentralized by avoiding central components. But while you have them, [securely manage your keys in the cloud](https://cloudsecurityalliance.org/research/topics/cloud-key-management/). 

> 2. **Centralized dapp control by the developer team:** Your project is not controlled by a DAO, for example because it is in an early development stage or does not (yet) hold significant funds. In that case, we recommend managing access to your canisters securely and ideally not letting individuals take full control of the application. To achieve that, consider requiring several individuals or parties to control canisters and to make security sensitive changes on the application level, such as modifying access controls, etc. A helpful tool to achieve this is the [threshold canister](https://github.com/dfinity/threshold). Manage key material using hardware security modules, such as e.g. [YubyHSM](https://www.yubico.com/ch/store/yubihsm-2-series/) and physically protect these. Some of these devices support threshold signature schemes, which can help to further secure the setup. 

Reviewer: I wonder if centralized = not controlled by DAO is true..
Where would blackholed canisters be in this picture? Are they considered centralized? I would say no.
I see this topic comes up later, but I wanted to record that when reaching this line I was wondering about that..

> 2. **Centralized dapp control by the developer team:** Your project is not controlled by a DAO, for example because it is in an early development stage or does not (yet) hold significant funds. In that case, we recommend managing access to your canisters securely and ideally not letting individuals take full control of the application. To achieve that, consider requiring several individuals or parties to control canisters and to make security sensitive changes on the application level, such as modifying access controls, etc. A helpful tool to achieve this is the [threshold canister](https://github.com/dfinity/threshold). Manage key material using hardware security modules, such as e.g. [YubyHSM](https://www.yubico.com/ch/store/yubihsm-2-series/) and physically protect these. Some of these devices support threshold signature schemes, which can help to further secure the setup. 

Reviewer: ```suggestion
2. **Centralized dapp control by the developer team:** Your project is not controlled by a DAO, for example because it is in an early development stage or does not (yet) hold significant funds. In that case, it is recommended to manage access to your canisters securely and ideally not letting individuals take full control of the application. To achieve that, consider requiring several individuals or parties to control canisters and to make security sensitive changes on the application level, such as modifying access controls, etc. A helpful tool to achieve this is the [threshold canister](https://github.com/dfinity/threshold). Manage key material using hardware security modules, such as e.g. [YubyHSM](https://www.yubico.com/ch/store/yubihsm-2-series/) and physically protect these. Some of these devices support threshold signature schemes, which can help to further secure the setup. 
```

I was not sure whether to read 
"consider requiring several individuals or parties (to control canisters and to make security sensitive changes on the application level)" or "(consider requiring several individuals or parties to control canisters) and to make security sensitive changes on the application level" and I was confused about what it means to "make security sensitive changes on the appliation level. Maybe this sentence can be simplified or split up?

Reviewer: Can also be resolved from my POV, thanks!

3. **Full decentralization using a DAO**: The dapp is controlled by a decentralized governance system such as the Internet Computer's [Service Nervous System (SNS)](../../developer-docs/integrations/sns), so that any security sensitive changes to the canisters are only executed if the SNS community approves them collectively through a proposal voting mechanism. If an SNS is used:
   - Make sure voting power is distributed over many, independent entities such that there is not one single or a few entities that can decide by themselves how the DAO evolves, see [here](../../developer-docs/integrations/sns/tokenomics/tokenomics-intro#voting-power-and-decentralization).

>    - Ensure all components of the dapp are under SNS control, including the canisters serving the web frontends. 

Reviewer: We could link here to 
how to use an asset canister under the SNS (https://internetcomputer.org/docs/current/developer-docs/integrations/sns/managing/sns-asset-canister)

   - Consider the [SNS preparation checklist](../../developer-docs/integrations/sns/tokenomics/sns-checklist). Important points from a security perspective are e.g. [tokenomics](../../developer-docs/integrations/sns/tokenomics/sns-checklist#11-tokenomics-specification), [disclosing dependencies to off-chain components](../../developer-docs/integrations/sns/tokenomics/sns-checklist#13-disclosure-of-dependencies) and [performing security reviews](../../developer-docs/integrations/sns/tokenomics/sns-checklist#21-security-review).

>    - Use an SNS on the SNS subnet as this guarantees that the SNS is running an NNS-blessed version and maintained as part of ICP.

Reviewer: ```suggestion
   - Rather than [self-deploying the SNS code](https://internetcomputer.org/docs/current/developer-docs/integrations/sns/introduction/dao-alternatives#self-deploy-the-sns-code), use the official SNS on the SNS subnet, as this guarantees that the SNS is running an NNS-blessed version and maintained as part of ICP.
```

(I am not sure if we should only call it an "SNS" if it is on the SNS subnet, so I thought this might make it a bit clearer. Or do you think this goes to far? )

Reviewer: After reading the last paragraph, I wonder to what extent we want to recommend this?
I think we should be open for alternative DAOs, but might also want to point out that the built-in SNS DAO comes with some security wins. I proposed to refrase it as "consider...". WDYT?

An alternative to DAO control (3. above) would be to create an immutable canister smart contract by removing the canister controller completely. However, note that this implies that the canister cannot be upgraded, which may have severe implications in case e.g. a bug were found. The option to use a decentralized governance system and thus being able to upgrade smart contracts is a big advantage of the Internet Computer ecosystem compared to other blockchains.

:::info
Note that, contrary to some other blockchains, also immutable smart contracts need cycles to run, and they can receive cycles.
:::

> It is also possible to implement a DAO [decentralized autonomous organization](https://en.wikipedia.org/wiki/Decentralized_autonomous_organization) on ICP from scratch. If you decide to do this (e.g. along the lines of the [basic DAO example](https://internetcomputer.org/docs/current/samples/dao)), be aware that this is security critical and must be security reviewed carefully. Furthermore, users will need to verify that the DAO is controlled by itself.

Reviewer: how about we link to these Wikis: https://wiki.internetcomputer.org/wiki/Verification_and_trust_in_a_(launched)_SNS  https://wiki.internetcomputer.org/wiki/SNS_decentralization_swap_trust
There was another one (here: https://wiki.internetcomputer.org/wiki/How-to:_Verify_SNS_decentralization_swap_proposal ) on how to verify a decentralization proposal, but since these proposals don't exist anymore, this has been removed.

> It is also possible to implement a DAO [decentralized autonomous organization](https://en.wikipedia.org/wiki/Decentralized_autonomous_organization) on ICP from scratch. If you decide to do this (e.g. along the lines of the [basic DAO example](https://internetcomputer.org/docs/current/samples/dao)), be aware that this is security critical and must be security reviewed carefully. Furthermore, users will need to verify that the DAO is controlled by itself.

Author: Excellent thanks @LaraAS and @roelstorms . Added these links in the section about SNS above.

> ### Verify the ownership of smart contracts you depend on

Author: @andrew-lee-work commented: change this to "decentralization" instead of ownership?

Reviewer: I agree, Or another alternative "control" (in the end it might be more important to know who controls than who "owns" it? although some might think of the controller when saying "own" but I think this has a less clear definition in the canister world)

#### Recommendations

In the following list, we first provide recommendations for centralized dapp control, and then move to recommendations for increasingly decentralized settings. From a security perspective, more decentralization is favorable. The following list could also be used as a basis for assessing a dapp's level of decentralization. This is just a set of recommendations and may not be complete. 
1. **The dapp uses central, off-chain components:** The application makes use of centralized components that are e.g. running in the cloud. The owners of these cloud services have full control over the application and assets managed by it. Your application should likely be further decentralized by avoiding central components. But while you have them, [securely manage your keys in the cloud](https://cloudsecurityalliance.org/research/topics/cloud-key-management/). 

> 2. **The dapp is controlled by the developer team:** Your project is not under decentralized control, for example because it is in an early development stage or does not (yet) hold significant funds. In that case, it is recommended to manage access to your canisters securely and ideally not letting individuals take full control of the application. To achieve that, consider requiring several individuals or parties to perform any security sentive operations. Such operations are e.g. canister controller operations, or security sensitive changes at the application level that are restricted to privileged principals, such as modifying access controls. A helpful tool to achieve this is the [threshold canister](https://github.com/dfinity/threshold). Manage key material using hardware security modules, such as e.g. [YubyHSM](https://www.yubico.com/ch/store/yubihsm-2-series/) and physically protect these. Some of these devices support threshold signature schemes, which can help to further secure the setup. To increase transparency about the changes made to a dapp, consider using a tool like [LaunchTrail](https://github.com/spinner-cash/launchtrail). 

Author: @domwoe : 

> @robin-kunzler It'd be great if the team could at some point have a look at https://github.com/spinner-cash/launchtrail since I think it is also an interesting approach between a dapp under centralized control and a DAO.

Thanks for the input, took a brief look at their readme and added a link here, see last sentence in this paragraph. WDYT?

Author: @domwoe I suppose this is OK and will resolve this thread so I can merge. If you have any input happy to take it!

Furthermore, the controller of canisters serving web content (such as e.g. the [asset canister](../../references/asset-canister)) could maliciously modify the web application to e.g. steal user funds or perform security sensitive actions on the user's behalf. For example, if [Internet Identity](../../developer-docs/integrations/internet-identity/overview) is used, the user principal's private key for the given origin is stored in the browser storage, and a malicious app can therefore fully control the private key and the user's session and any assets controlled by that key.  

Dapps are commonly reachable over their own custom domain name instead of ic0.app. These domains are registered with a DNS registrar by one of the developers. The developer can choose to have this domain point at a completely different web application. Even one not hosted on the IC. Users will trust this domain and the app it serves. This could allow such a developer to steal funds, leak data, etc.

> A dApp might have privileged features which are only accessible to principals that are on an allow list. For example, minting new tokens, debug functions, permissions management, removing NFTs for digital rights violations, etc. This means that whoever controls that principal (this could e.g. be dapp developers) may have central control over these privileged features.

Reviewer: ```suggestion
A dapp might have privileged features which are only accessible to principals that are on an allow list. For example, minting new tokens, debug functions, permissions management, removing NFTs for digital rights violations, etc. This means that whoever controls that principal (this could e.g. be dapp developers) may have central control over these privileged features.
```

Author: Thanks, done!
